src/curve/LinearIncreasingCurve.sol

```diff
@@ -380,42 +386,75 @@ contract LinearIncreasingEscrow is
+                if (oldDSlope < 0) oldDSlope = 0;
+                slopeChanges[_fromLockedEnd] = oldDSlope;
             }
         }
 
         if (lastPoint.slope < 0) lastPoint.slope = 0;
         if (lastPoint.bias < 0) lastPoint.bias = 0;
+        if (newDSlope < 0) newDSlope = 0;
 
-        // TODO: see aerodome..
-        globalPointLatestIndex = _globalPointLatestIndex;
-        _globalPointHistory[_globalPointLatestIndex] = lastPoint;
+        // store new slope change
+        slopeChanges[newEnd] = newDSlope;
 
-        slopeChanges[newEnd] = newDSlope < 0 ? int256(0) : newDSlope;
+        // Record the latest global point.
+        _storeLatestGlobalPoint(lastPoint, _globalPointLatestIndex);
 
         // Create new token point and store.
         TokenPoint memory tNew;
         tNew.writtenTs = uint128(block.timestamp);
         tNew.checkpointTs = _newLocked.start;
         tNew.coefficients = [newLockBias, newLockSlope, 0];
 
-        if (
-            tokenLatestIndex != 0 &&
-            _tokenPointHistory[_tokenId][tokenLatestIndex].writtenTs == block.timestamp
-        ) {
-            _tokenPointHistory[_tokenId][tokenLatestIndex] = tNew;
+        // Record the latest token point.
+        _storeLatestTokenPoint(tNew, _tokenId, tokenLatestIndex);
+    }
+
+    /// @dev The private helper function to either store latest global point on a new index or overwrite it.
+    ///      In case of overwriting, the latest global point index is not incremented.
+    function _storeLatestGlobalPoint(GlobalPoint memory _p, uint256 _index) private {
+        // If the timestamp of last stored global point is the same as
+        // current timestamp, overwrite it, otherwise store a new one
+        // to reduce unnecessary global points in the history for
+        // gas costs and binary search efficiency.
+        if (_index != 1 && _globalPointHistory[_index - 1].writtenTs == block.timestamp) {
+            _globalPointHistory[_index - 1] = _p;
         } else {
-            tokenPointLatestIndex[_tokenId] = ++tokenLatestIndex;
-            _tokenPointHistory[_tokenId][tokenLatestIndex] = tNew;
+            globalPointLatestIndex = _index;
+            _globalPointHistory[_index] = _p;
+        }
+    }
+
+    /// @dev The private helper function to either store latest token point on a new index or overwrite it.
+    ///      In case of overwriting, the latest token point index is not incremented.
+    function _storeLatestTokenPoint(
+        TokenPoint memory _p,
+        uint256 _tokenId,
+        uint256 _index
+    ) private {
+        // If the timestamp of last stored token point is the same as
+        // current timestamp, overwrite it, otherwise store a new one
+        // to reduce unnecessary global points in the history for
+        // gas costs and binary search efficiency.
+        if (_index != 0 && _tokenPointHistory[_tokenId][_index].writtenTs == block.timestamp) {
```

Comment:
IIRC tokenPointHistory uses index 0 as an heuristic for empty array. 

Why then are we differentiating behaviours? 

Put another way - why does Index 1 indicate the start of global points, but index 0 the start of TokenPoints?

Reply:
This has to do with the main loop inside `_checkpoint`. Note that inside loop, we increase globalPointindex by one. meaning that after loop, it becomes 1. That's not the same situation for token points as they don't have loop. What you suggest can be done, but that would require extra if inside the main loop and make things hard to debug.